tokio/src/sync/mpsc/chan.rs

```diff
@@ -293,6 +295,82 @@ impl<T, S: Semaphore> Rx<T, S> {
         })
     }
 
+    /// Receives values into `buffer` up to its capacity
+    ///
+    /// If the buffer 0 remaining capacity but a value is received, reserves `super::BLOCK_CAP` elements
+    pub(crate) fn recv_many(&mut self, cx: &mut Context<'_>, buffer: &mut Vec<T>) -> Poll<usize> {
+        use super::block::Read;
+
+        let mut insufficient_capacity = buffer.capacity() == buffer.len();
+        let initial_length = buffer.len();
+
+        ready!(crate::trace::trace_leaf(cx));
+        // Keep track of task budget
+        let coop = ready!(crate::runtime::coop::poll_proceed(cx));
+
+        self.inner.rx_fields.with_mut(|rx_fields_ptr| {
+            let rx_fields = unsafe { &mut *rx_fields_ptr };
+            macro_rules! try_recv {
+                () => {
+                    while (buffer.len() < buffer.capacity() || insufficient_capacity) {
+                        match rx_fields.list.pop(&self.inner.tx) {
+                            Some(Read::Value(value)) => {
+                                if insufficient_capacity {
+                                    buffer.reserve(super::BLOCK_CAP);
+                                    insufficient_capacity = false;
+                                }
+                                buffer.push(value);
+                            }
+
+                            Some(Read::Closed) => {
+                                let number_added = buffer.len() - initial_length;
+                                if number_added > 0 {
+                                    self.inner.semaphore.add_permits(number_added);
+                                }
+                                // TODO: This check may not be required as it most
+                                // likely can only return `true` at this point. A
+                                // channel is closed when all tx handles are
+                                // dropped. Dropping a tx handle releases memory,
+                                // which ensures that if dropping the tx handle is
+                                // visible, then all messages sent are also visible.
+                                assert!(self.inner.semaphore.is_idle());
+                                coop.made_progress();
+                                return Ready(number_added);
+                            }
+
+                            None => {
+                                break; // fall through
+                            }
+                        }
+                    }
+                    let number_added = buffer.len() - initial_length;
+                    if number_added > 0 {
+                        self.inner.semaphore.add_permits(number_added);
+                        coop.made_progress();
+                        return Ready(number_added);
+                    }
+                };
+            }
```

Comment:
This is getting complicated again. Let's try to simplify a bit. How about this?

 * At the beginning instead of `insufficient_capacity`, define a constant called `max_length` set to `buffer.capacity()`, or `buffer.len() + BLOCK_CAP` when we want to resize.
 * In the while loop, just use `buffer.len() < max_length`.
 * We don't resize it explicitly. Instead, we just call `Vec::push`, which will resize for us.

Reply:
Reposted my email reply in this thread:

> My concern is on the second call to after the push that effectively reserves
> the buffer.  As capacity grows exponentially on push, then naive use of
> recv_many eventually results in taking in all elements in future calls.
> (So why initially even limit max_length to buffer.len() + BLOCK_CAP?)

_I promise to not make a habit of replying from email._

I suggest the following use cases:

1. The caller clears the buffer between calls to `recv_many`; capacity sets the number of messages reasonable to process in one call so as to not affect liveness.
2.  The caller can anticipate the total number of messages they will receive and want to pre-allocate that space by setting the capacity explicitly; liveness is known to not be a concern.
3. The caller wants to indefinitely extend the buffer.  It makes sense to the buffer's capacity grows exponentially, e.g., how push does it.
  i. The caller knows that the rate of sent messages on the channel will not affect liveness no matter how many are retrieved.
  ii. The caller needs to limit how many messages are added to the buffer at a time to ensure liveness.

Use-case 3 seems to be in conflict with use-case 1.  It seems clear that one possible (maybe the only) resolution is to pass in an argument to limit the number of messages retrieved.  Furthermore, passing in a bound to the number of retrieved messages helps generalize case 2.

Thoughts?

Reply:
It's not clear to me how the use-cases are in conflict, but I am also ok with using an argument to limit the number of messages if you prefer that.

Reply:
I assume there is a very busy sender.

If the buffer holds say, 1,000,0000 elements and has no unused capacity, upon the call to `recv_many` the capacity is doubled and `BLOCK_CAP` (few) values are retrieved. On the _next_ call, `recv_many` will  populate 1,000,0000 - `BLOCK_CAP` elements in a single call (as necessitated by how use-case #1 works -- which allows up to the capacity to be retrieved).  In the next call, `recv_many` finds the buffer full and doubles it again, and `BLOCK_CAP` elements are retrieved.  On the next call to `recv_many` -- 2,000,000 - BLOCK_CAP elements are retrieved in a single call.

To me this is a very odd behavior where between calls it oscillates between returning a small number of values (on the step where capacity is doubled) and an exponentially increasing number of values (because of significant unused capacity.)  

This fails to satisfy use-case 3.ii -- where the "caller needs to limit how many messages are added to the buffer at a time to ensure liveness."

This was why I added constant capacity on each call -- so that naive use of `recv_many` only resulted in, at most, a fixed number of values being added to the buffer with each `recv_many` call.  (And why an earlier commit simply panicked if there was no spare capacity.)

Reply:
I assume there is unlimited flood of messages being sent.  The problem arises when a `Vec::push` adds significant unused capacity to the buffer and the behavior of use-case 1 takes over, resulting in a potential huge number of messages being put into the buffer at once, violating 3.ii -- too many messages at once can make things unresponsive. The problem becomes worse as the buffer grows. 

To resolve this -- I have re-worked to pass in a limit to the number of received values.

Reply:
I guess that makes sense.